### PR TITLE
Add `Sum` aggregation function 

### LINKED ISSUE #2892: Support for `sum` aggregation function while summarizing columns
## Problem
Currently, two aggregation functions (**list & count**) are supported as transformation steps while summarizing columns. Besides, there are many useful aggregation functions and `sum` is one of them.

## Proposed solution
It would be nice to support `sum` as a transformation step while summarizing columns.

## To Discuss
- `sum` aggregation function should only be available for columns with **Number** data-type (and **Boolean**?).

## Additional context
This feature request is a part of GSoC '23 project, [Add More Summarization Functions](https://summerofcode.withgoogle.com/programs/2023/projects/D9XvkQCN).

### PR DESCRIPTION
**Pull request by Aritra8438**
Fixes #2892 

**This pull request adds support for `Sum` aggregation as a transformation step in the data explorer.** 

User will be able to:
- Get summation of integer columns based on `Group By` clause.
- Get summation of decimal columns based on `Group By` clause.
---
**Technical details**

I have made the following changes:
- Add `Sum(DBFunction)` class to support sum aggregation
- Add `sum` suffix to the output alias of the aggregated column.
- Add `Sum` to the list of transformation steps in the data explorer.

---
**Screencast**

https://github.com/centerofci/mathesar/assets/64671908/62e6e419-6731-457f-974e-df6b33ced3eb

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/functions/base.py
Comment (mathemancer) [inline db/functions/base.py:408]: This is not accurate w.r.t. the return type, but it doesn't seem to affect the front end.

I suppose we can leave it this way since these return type declarations are going to be removed as we remove SQL Alchemy.

Comment (mathemancer) [inline db/functions/base.py:408]: This is not accurate w.r.t. the return type, but it doesn't seem to affect the front end.

I suppose we can leave it this way since these return type declarations are going to be removed as we remove SQL Alchemy.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Overall, this is pretty good from a back end perspective.

The only change I have to request is that I'd like you to add a test or two in
`mathesar/tests/api/query/test_library_demo_reports.py` showing off the sum summarize function.

We should consider breaking out a separate API layer test suite for testing summarizations as we add more and more.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Overall, this is pretty good from a back end perspective.

The only change I have to request is that I'd like you to add a test or two in
`mathesar/tests/api/query/test_library_demo_reports.py` showing off the sum summarize function.

We should consider breaking out a separate API layer test suite for testing summarizations as we add more and more.

**Comment by mathemancer** [comment]
I'm going to request a review from @seancolsen since there's some front end work on this.

**Comment by Aritra8438** [comment]
Sure, @mathemancer. 

The `sum` aggregation function should only be available for `numeric` columns. 
(In the latest commit) I have modified the data explorer UI so that only available aggregation functions are rendered to the users. (i.e. **for text columns, users won't have the option to select the sum as an aggregation function**)

**Review by mathemancer — CHANGES_REQUESTED** [review]
Let's move the aggregation test to a new file:
`mathesar/tests/api/query/test_aggregations`.

Sorry for the mixed messages.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Let's move the aggregation test to a new file:
`mathesar/tests/api/query/test_aggregations`.

Sorry for the mixed messages.

**Comment by Aritra8438** [comment]
@mathemancer, I have moved the aggregation-functions tests to `mathesar/tests/api/query/test_aggregation_functions.py`.

I have broken down the tests to aggregation level to track `specs` for each aggregation.

**Review by mathemancer — APPROVED** [review]
Okay, this looks good to me. Thank you!

**Review by mathemancer — APPROVED** [review]
Okay, this looks good to me. Thank you!

**Comment by mathemancer** [comment]
@Aritra8438 Please wait for review from @seancolsen to review.

**Comment by seancolsen** [comment]
@Aritra8438 and @mathemancer I want to give you a heads up that I'm working on this.

but there are some substantial front end changes, and it's taking me a little time.

**Review by seancolsen — APPROVED** [review]

**Review by pavish — APPROVED** [review]
@seancolsen This looks good!

> I don't think it's quite done yet. In our [most recent conversation](https://matrix.to/#/!UnujZDUxGuMrYdvgTU:matrix.mathesar.org/$gy6AfE-G-557O6CCSSoU7KVs_IgzEsMGqPnlrQ7o8f0?via=matrix.mathesar.org&via=matrix.org) you recommended making some changes within speculateColumnMetaData. I didn't make those changes.

I think this can be safely done later.

**Pull request merged**
